internal/engine/runtimelog/podlogmanager_test.go

```diff
@@ -228,6 +227,82 @@ func TestTerminatedContainerLogs(t *testing.T) {
 	f.AssertOutputDoesNotContain("goodbye world!")
 }
 
+// https://github.com/tilt-dev/tilt/issues/3908
+func TestLogReconnection(t *testing.T) {
+	f := newPLMFixture(t)
+	defer f.TearDown()
+
+	state := f.store.LockMutableStateForTesting()
+
+	cName := container.Name("cName")
+	p := store.Pod{
+		PodID: podID,
+		Containers: []store.Container{
+			NewRunningContainer(cName, "cID"),
+		},
+	}
+	state.UpsertManifestTarget(manifestutils.NewManifestTargetWithPod(
+		model.Manifest{Name: "server"}, p))
+	f.store.UnlockMutableState()
+
+	reader, writer := io.Pipe()
+	defer writer.Close()
+	f.kClient.SetLogReaderForPodContainer(podID, cName, reader)
+
+	// Set up fake time
+	startTime := time.Now()
+	currentTime := startTime.Add(5 * time.Second)
+	timeCh := make(chan time.Time)
+	ticker := time.Ticker{C: timeCh}
+	f.plm.now = func() time.Time { return currentTime }
+	f.plm.since = func(t time.Time) time.Duration { return currentTime.Sub(t) }
+	f.plm.newTicker = func(d time.Duration) *time.Ticker { return &ticker }
+
+	f.store.WithState(func(state *store.EngineState) {
+		state.TiltStartTime = startTime
+	})
+
+	f.plm.OnChange(f.ctx, f.store)
+
+	_, _ = writer.Write([]byte("hello world!"))
+	f.AssertOutputContains("hello world!")
+	assert.Equal(t, startTime, f.kClient.LastPodLogStartTime)
+
+	currentTime = currentTime.Add(20 * time.Second)
+	lastRead := currentTime
+	_, _ = writer.Write([]byte("hello world2!"))
+	f.AssertOutputContains("hello world2!")
+
+	// Simulate Kubernetes rotating the logs by creating a new pipe.
+	reader2, writer2 := io.Pipe()
+	defer writer2.Close()
+	f.kClient.SetLogReaderForPodContainer(podID, cName, reader2)
+	go func() {
+		_, _ = writer2.Write([]byte("goodbye world!"))
+	}()
+	f.AssertOutputDoesNotContain("goodbye world!")
+
+	currentTime = currentTime.Add(5 * time.Second)
+	timeCh <- currentTime
+	f.AssertOutputDoesNotContain("goodbye world!")
+
+	currentTime = currentTime.Add(5 * time.Second)
+	timeCh <- currentTime
+	f.AssertOutputDoesNotContain("goodbye world!")
+	assert.Equal(t, startTime, f.kClient.LastPodLogStartTime)
+
+	currentTime = currentTime.Add(5 * time.Second)
```

Comment:
maybe worth a comment: "simulate 15s since we last read a log; this triggers a reconnect"

Reply:
done